shared/origin.js

```diff
@@ -0,0 +1,183 @@
+			"rememberMe": "on"
+		}
+		let authResponse = await PostData( jSessionLocation, authData, {"Cookie": [`JSESSIONID=${jSessionID}`, `signin-cookie=${signinCookie}`]} )
+		let authResLocation = authResponse.toString().match( /(?<=window\.location = ")\S+(?=";)/g )[0]
+		let sidHeaders = await GetHeaders( authResLocation, {"Cookie": [`JSESSIONID=${jSessionID}`, `signin-cookie=${signinCookie}`]} )
+		sidCookie = sidHeaders["set-cookie"].join( "; " ).match( /(?<=sid=)[\S]+?(?=;)/g )[0]
+	}
+	let authTokenRes = await GetData( loginURL, {"Cookie": [`sid=${sidCookie}`]} )
+	let authResJson = JSON.parse( authTokenRes.toString() )
+	AuthToken = authResJson.access_token
+	console.log( "Successfully got Origin auth token" )
+	authed = true
+
+	setTimeout( authWithOrigin, Number( authResJson.expires_in )*1000 - 60000 ) // Refresh access token 1 minute before it expires just to be safe
+}
+
+authWithOrigin()
+
+function GenerateCID()
+{
+	var l = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXTZabcdefghiklmnopqrstuvwxyz"
+	var h = 32
+	var j = ""
+	for ( var k = 0; k < h; k++ )
+	{
+		var m = Math.floor( Math.random() * l.length )
+		j += l.substring( m, m + 1 )
+	}
+	return j
+}
+
+function GetHeaders( location, headers = {} )
+{
+	return new Promise( resolve =>
+	{
+		let params = { headers }
+		let href = new URL( location )
+		params.host = href.host
+		params.path = href.pathname + href.search
+		https.get( params, reqResult =>
+		{
+			resolve( reqResult.headers )
+		} )
+	} )
+}
+function GetData( location, headers = {} )
+{
+	return new Promise( resolve =>
+	{
+		let params = { headers }
+		let href = new URL( location )
+		params.host = href.host
+		params.path = href.pathname + href.search
+		https.get( params, reqResult =>
+		{
+			let data = []
+			reqResult.on( "data", c => data.push( c ) )
+			// eslint-disable-next-line
+			reqResult.on( "end", _ => resolve( Buffer.concat( data ) ) )
+		} )
+	} )
+}
+
+// data must be an object to be sent as x-www-form-urlencoded
+function PostData( location, postData, headers = {} )
+{
+	return new Promise( ( resolve, reject ) =>
+	{
+		let params = { headers }
+		let href = new URL( location )
+		params.method = "POST"
+
+		params.host = href.host
+		params.path = href.pathname + href.search
+
+		if ( postData )
+		{
+			var body = new URLSearchParams()
+			for( var name in postData )
+			{
+				body.append( name, postData[name] )
+			}
+
+			headers["Content-Length"] = body.toString().length
+			headers["Content-Type"] = "application/x-www-form-urlencoded"
+		}
+		let req = https.request( params, reqResult =>
+		{
+			let data = []
+			reqResult.on( "data", c => data.push( c ) )
+			// eslint-disable-next-line
+			reqResult.on( "end", _ => resolve( Buffer.concat( data ) ) )
+		} )
+
+		req.on( "error", e =>
+		{
+			reject( e )
+		} )
+
+		if ( postData )
+		{
+			req.write( body.toString() )
+		}
+
+		req.end()
+	} )
+}
+
+const asyncHttp = require( "./asynchttp.js" )
+
+function getUserInfo( uid )
```

Comment:
Why is this not just an async function? The resolve/reject stuff inside makes it overly complicated.